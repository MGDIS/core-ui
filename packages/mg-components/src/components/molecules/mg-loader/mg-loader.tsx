import { Component, Element, h, Prop, Watch } from '@stencil/core';
import { initLocales } from '../../../locales';
import { isValidString } from '@mgdis/stencil-helpers';

@Component({
  tag: 'mg-loader',
  styleUrl: '../../../../node_modules/@mgdis/styles/dist/components/mg-loader.css',
  shadow: true,
})
export class MgLoader {
  /************
   * Internal *
   ************/

  // Locales
  private messages;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgLoaderElement;

  /**
   * Override loader message
   */
  @Prop() message?: string;
  @Watch('message')
  watchMessage(newValue: MgLoader['message']): void {
    if (newValue && !isValidString(newValue)) throw new Error('<mg-loader> prop "message" must be a valid string.');
  }

  /**
   * Hide message
   */
  @Prop() messageHide?: boolean;

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   */
  componentWillLoad(): void {
    // Get locales
    this.messages = initLocales(this.element).messages;

    // validate props
    this.watchMessage(this.message);
  }

  /**
   * Render
   * @returns HTML Element
   */
  render(): HTMLElement {
    return (
      <div class="mg-c-loader" aria-live="polite">
        <mg-icon icon="loader" spin></mg-icon>
        <span class={{ 'mg-u-visually-hidden': this.messageHide }}>{this.message || this.messages.loader.inProgess}</span>
      </div>
    );
  }
}
