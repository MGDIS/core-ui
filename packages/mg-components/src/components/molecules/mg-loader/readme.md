## Specifications

![](./doc/img/mg-loader-anatomy.png)

### Anatomy

Component is built with :

- loader icon
- loader message

### Style

message : Open Sans, 13px, Regular
All contents are centered.

<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description             | Type      | Default     |
| ------------- | -------------- | ----------------------- | --------- | ----------- |
| `message`     | `message`      | Override loader message | `string`  | `undefined` |
| `messageHide` | `message-hide` | Hide message            | `boolean` | `undefined` |


## Dependencies

### Depends on

- [mg-icon](../../atoms/mg-icon)

### Graph
```mermaid
graph TD;
  mg-loader --> mg-icon
  style mg-loader fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
