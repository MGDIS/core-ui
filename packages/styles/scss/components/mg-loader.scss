@import '../utils/a11y';

:host {
    display: block;
}

.mg-c-loader {
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    justify-content: center;
    gap: 0.4rem;
}