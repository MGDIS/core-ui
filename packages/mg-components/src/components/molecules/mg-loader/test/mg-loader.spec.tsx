import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgLoader } from '../mg-loader';

const getPage = (args: Partial<Pick<MgLoader, 'message' | 'messageHide'> & { lang: string }>) =>
  newSpecPage({
    components: [MgLoader],
    template: () => <mg-loader {...args}></mg-loader>,
  });

describe('mg-loader', () => {
  test.each([{}, { message: 'new loader message' }, { messageHide: true }])('with args %s', async args => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test('Should throw error if message is not valid', async () => {
    expect.assertions(1);
    try {
      await getPage({ message: ' ' });
    } catch (err) {
      expect(err.message).toEqual('<mg-loader> prop "message" must be a valid string.');
    }
  });

  test.each(['fr', 'xx'])('Should render component with locale: %s', async lang => {
    const { root } = await getPage({ lang });
    expect(root).toMatchSnapshot();
  });
});
