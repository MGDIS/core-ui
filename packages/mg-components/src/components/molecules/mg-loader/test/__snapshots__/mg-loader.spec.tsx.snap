// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-loader Should render component with locale: fr 1`] = `
<mg-loader lang="fr">
  <mock:shadow-root>
    <div aria-live="polite" class="mg-c-loader">
      <mg-icon icon="loader" spin=""></mg-icon>
      <span>
        Chargement en cours...
      </span>
    </div>
  </mock:shadow-root>
</mg-loader>
`;

exports[`mg-loader Should render component with locale: xx 1`] = `
<mg-loader lang="xx">
  <mock:shadow-root>
    <div aria-live="polite" class="mg-c-loader">
      <mg-icon icon="loader" spin=""></mg-icon>
      <span>
        Loading in progress...
      </span>
    </div>
  </mock:shadow-root>
</mg-loader>
`;

exports[`mg-loader with args { message: 'new loader message' } 1`] = `
<mg-loader>
  <mock:shadow-root>
    <div aria-live="polite" class="mg-c-loader">
      <mg-icon icon="loader" spin=""></mg-icon>
      <span>
        new loader message
      </span>
    </div>
  </mock:shadow-root>
</mg-loader>
`;

exports[`mg-loader with args { messageHide: true } 1`] = `
<mg-loader>
  <mock:shadow-root>
    <div aria-live="polite" class="mg-c-loader">
      <mg-icon icon="loader" spin=""></mg-icon>
      <span class="mg-u-visually-hidden">
        Loading in progress...
      </span>
    </div>
  </mock:shadow-root>
</mg-loader>
`;

exports[`mg-loader with args {} 1`] = `
<mg-loader>
  <mock:shadow-root>
    <div aria-live="polite" class="mg-c-loader">
      <mg-icon icon="loader" spin=""></mg-icon>
      <span>
        Loading in progress...
      </span>
    </div>
  </mock:shadow-root>
</mg-loader>
`;
